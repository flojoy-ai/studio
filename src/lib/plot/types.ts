export interface Drawable {
  draw: () => void;
}
