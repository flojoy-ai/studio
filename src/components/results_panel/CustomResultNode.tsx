import { Handle, Position } from "react-flow-renderer";
import { useFlowChartState } from "../../hooks/useFlowChartState";
import Plot from "react-plotly.js";
import styledPlotLayout from "./../defaultPlotLayout";
import PlotlyComponent from "../plotly-wrapper/PlotlyComponent";

const CustomResultNode = ({ data }) => {
  const { uiTheme } = useFlowChartState();
  const styledLayout = styledPlotLayout(uiTheme);
  return (
    <div style={{ position: "relative" }}>
      {(data.func === "MULTIPLY" || data.func === "ADD") && (
        <>
          <Handle
            type="target"
            position={Position.Left}
            style={{ borderRadius: 0 }}
          />
          <Handle
            type="source"
            position={Position.Right}
            style={{ borderRadius: 0 }}
          />
        </>
      )}
      {data.type === "VISOR" && (
        <>
          <Handle
            type="target"
            position={Position.Left}
            style={{ borderRadius: 0 }}
          />
          <Handle
            type="source"
            position={Position.Right}
            style={{ borderRadius: 0 }}
          />
        </>
      )}
      {!(data.func === "MULTIPLY" || data.func === "ADD") &&
        data.type !== "VISOR" && (
          <>
            <Handle
              type="target"
              position={Position.Left}
              style={{ borderRadius: 0 }}
            />
            <Handle
              type="source"
              position={Position.Right}
              style={{ borderRadius: 0 }}
            />
          </>
        )}

      {data?.resultData !== null ? (
        <PlotlyComponent
          id={data.func}
          data={
            "data" in data.resultData
              ? data.resultData.data
              : [{ x: data.resultData["x0"], y: data.resultData["y0"] }]
          }
          layout={
            "layout" in data.resultData
              ? Object.assign({}, data.resultData.layout, styledLayout)
              : Object.assign({}, { title: `${data.func}` }, styledLayout)
          }
          useResizeHandler
          style={{
            height: 235,
            width: 230,
          }}
        />
      ) : (
        <p> `NO Result`</p>
      )}
    </div>
  );
};

export default CustomResultNode;
