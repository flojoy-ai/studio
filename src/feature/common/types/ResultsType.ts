import { OverridePlotData } from "@/types";
import { Layout } from "plotly.js";

export type ResultIO = {
  cmd: string;
  id: string;
  result: Result;
};
export type NodeResult = {
  cmd: string;
  id: string;
  result: Result;
};

export type Result = {
  plotly_fig?: {
    data: OverridePlotData;
    layout: Partial<Layout> | undefined;
  };
  text_blob?: string;
  data?: ScalarData | OrderedPairData;
};

export type ScalarData = {
  c: number;
};

export type OrderedPairData = {
  x: number[];
  y: number[];
};
