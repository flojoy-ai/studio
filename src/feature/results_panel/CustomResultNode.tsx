import { Handle, Position } from "react-flow-renderer";
import { useFlowChartState } from "../../hooks/useFlowChartState";
import Plot from "react-plotly.js";
import styledPlotLayout from "../../components/defaultPlotLayout";

const CustomResultNode = ({ data }) => {
  const { uiTheme } = useFlowChartState();
  const styledLayout = styledPlotLayout(uiTheme);
  
  return (
    <div style={{ position: "relative" }} data-testid="result-node">
      <>
        <Handle
          type="target"
          position={Position.Left}
          style={{ borderRadius: 0 }}
        />
        <Handle
          type="source"
          position={Position.Right}
          style={{ borderRadius: 0 }}
        />
      </>

      {!data?.resultData ? (
        <p> `NO Result`</p>
      ) : (
        <Plot
          data={
            !data.resultData?.data
              ? [{ x: data.resultData["x"], y: data.resultData["y"] }]
              : data.resultData.data
          }
          layout={
            !data.resultData?.layout
              ? Object.assign({}, { title: `${data.func}` }, styledLayout)
              : Object.assign({}, data.resultData.layout, styledLayout)
          }
          useResizeHandler
          style={{
            height: 235,
            width: 230,
          }}
        />
      )}
    </div>
  );
};

export default CustomResultNode;
