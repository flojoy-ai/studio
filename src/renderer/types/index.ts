export type {
  ALL_DC_TYPE,
  DataContainer,
  DataContainerType,
} from "./dataContainer";
export type { CustomNodeProps, BlockData } from "./node";
export type {
  DataContainer2PlotlyProps,
  OverridePlotData,
  PlotProps,
} from "./plotly";
