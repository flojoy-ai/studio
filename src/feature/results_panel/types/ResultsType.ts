import { ElementsData } from "@src/feature/flow_chart_panel/types/CustomNodeProps";
import { Data, Layout } from "plotly.js";

export type ResultIO = {
  cmd: string;
  id: string;
  result: {
    default_fig: {
      data: Data[],
      layout: Partial<Layout>
    },
    data: {
      type: string;
      [key:string]: any;
    }
  };
};
export interface ResultsType {
  io?: ResultIO[];
}
export type ResultNodeData = ElementsData & {
  resultData?: ResultIO["result"];
};
