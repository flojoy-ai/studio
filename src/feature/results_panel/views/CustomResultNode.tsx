import { useMantineTheme } from "@mantine/core";
import PlotlyComponent from "@src/feature/common/PlotlyComponent";
import usePlotLayout from "@src/feature/common/usePlotLayout";
import { Handle, Position } from "reactflow";
import { ResultNodeData } from "../types/ResultsType";
import { makePlotlyData } from "@src/utils/format_plotly_data";

type CustomResultNodeProp = {
  data: ResultNodeData;
};

const CustomResultNode = ({ data }: CustomResultNodeProp) => {
  const theme = useMantineTheme();
  const styledLayout = usePlotLayout();

  return (
    <div style={{ position: "relative" }} data-testid="result-node">
      <>
        <Handle
          type="target"
          position={Position.Left}
          style={{ borderRadius: 0 }}
        />
        <Handle
          type="source"
          position={Position.Right}
          style={{ borderRadius: 0 }}
        />
      </>

      {!data?.resultData ? (
        <p> NO Result </p>
      ) : (
        <PlotlyComponent
          id={data.id}
          data={makePlotlyData(data.resultData.default_fig.data, theme)}
          layout={Object.assign({}, { title: data.label }, styledLayout)}
          useResizeHandler
          style={{
            height: 190,
            width: 210,
          }}
        />
      )}
    </div>
  );
};

export default CustomResultNode;
