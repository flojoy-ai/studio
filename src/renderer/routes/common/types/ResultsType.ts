import { OverridePlotData } from "@/renderer/types";
import { Layout } from "plotly.js";

export type ResultIO = {
  cmd: string;
  id: string;
  result: Result;
};
export type BlockResult = {
  cmd: string;
  id: string;
  result: Result;
};

export type Result = {
  plotly_fig?: {
    data: OverridePlotData;
    layout: Partial<Layout> | undefined;
  };
  text_blob?: string;
  data?: ScalarData;
};

export type ScalarData = {
  c: number;
};
