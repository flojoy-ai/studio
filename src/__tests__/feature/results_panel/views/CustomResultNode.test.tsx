import { render } from "@testing-library/react";
import CustomResultNode from "@src/feature/results_panel/views/CustomResultNode";
import { ResultNodeData } from "@src/feature/results_panel/types/ResultsType";

jest.mock("@src/hooks/useFlowChartState", () => ({
  useFlowChartState: jest.fn(() => ({ uiTheme: "dark" })),
}));
jest.mock("reactflow", () => {
  const Handle = jest.fn().mockReturnValue(<div>Handle Component</div>);
  const Position = { Left: "LEFT", Right: "RIGHT" };
  return {
    Handle,
    Position,
  };
});
jest.mock("@src/feature/common/PlotlyComponent", () => ({
  __esModule: true,
  default: jest.fn((props) => (
    <div
      id={props.id}
      style={props.style}
      data-imgsource={
        props.data.length > 0 && props.data[0].source && props.data[0].source
      }
    >
      PlotlyComponent
    </div>
  )),
}));
describe("CustomResultNode", () => {
  const data: ResultNodeData = {
    id: "node-id",
    func: "some function",
    ctrls: {},
    label: "node-label",
    type: "SINE",
    resultData: {
      type: "bar",
      x: [1, 2, 3],
      y: [4, 5, 6],
      layout: {
        title: "some title",
      },
    },
  };
  it("renders a PlotlyComponent if result data is provided", () => {
    const { container, getByTestId } = render(<CustomResultNode data={data} />);
    const resultNode = getByTestId("result-node");
    expect(resultNode).toBeInTheDocument();
    const plotlyComponent = resultNode.querySelector(`#${data.id}`);
    expect(plotlyComponent).toBeInTheDocument();
    expect(plotlyComponent).toHaveStyle({ height: "190px", width: "210px" });
    expect(container).toMatchSnapshot();
  });

  it("renders a 'NO Result' message if no result data is provided", () => {
    const { getByText } = render(
      <CustomResultNode data={{ ...data, resultData: undefined }} />
    );
    expect(getByText("NO Result")).toBeInTheDocument();
  });

  it("converts image result data to a data URL", () => {
    const modifiedData: ResultNodeData = {
      ...data,
      resultData: {
        type: "image",
        y: [[1, 2, 3]],
        file_type: ["image"],
      },
    };
    const { container, getByTestId } = render(
      <CustomResultNode data={modifiedData} />
    );
    const resultNode = getByTestId("result-node");
    const plotlyComponent = resultNode.querySelector(`#${modifiedData.id}`);
    expect(plotlyComponent).toBeInTheDocument();
    expect(container).toMatchSnapshot("__image_type_file__");
    expect(plotlyComponent).toHaveAttribute(
      "data-imgsource",
      "data:image/jpeg;base64," +
        Buffer.from(
          (modifiedData.resultData?.y![0] as number[]) || []
        ).toString("base64")
    );
  });
});
