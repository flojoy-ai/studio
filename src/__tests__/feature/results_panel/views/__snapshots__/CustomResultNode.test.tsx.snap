// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomResultNode renders a PlotlyComponent if result data is provided 1`] = `
<div>
  <div
    data-testid="result-node"
    style="position: relative;"
  >
    <div>
      Handle Component
    </div>
    <div>
      Handle Component
    </div>
    <div
      id="node-id"
      style="height: 293px; width: 380px;"
    >
      PlotlyComponent
    </div>
  </div>
  ,
</div>
`;
